src/common/types/vector.cpp

```diff
@@ -709,16 +709,23 @@ void Vector::Sequence(int64_t start, int64_t increment) {
 
 void Vector::Serialize(idx_t count, Serializer &serializer) {
 	auto &type = GetType();
+
+	VectorData vdata;
+	Orrify(count, vdata);
+
+	const auto write_validity = (count > 0) && !vdata.validity.AllValid();
+	serializer.Write<bool>(write_validity);
+	if (write_validity) {
+		serializer.WriteData((const_data_ptr_t)vdata.validity.GetData(), vdata.validity.ValidityMaskSize(count));
```

Comment:
This will likely not work if you are serializing a dictionary or constant vector, as you are serializing the validity mask as if the vector was uncompressed.

I believe the vector serialization is only used for serializing inserts/updates/deletes in the WAL currently, and those are likely all flat vectors, hence this will not be triggered by the tests. 

The fix would be to either add a call to `Normalify` to turn the vector into a flat uncompressed vector before doing this operation, or to use the selection vector (`vdata.sel`) to write the validity mask.

Reply:
Ah so like in `CopyToStorageLoop`? I had wondered about the point of the `Vector`s built in selection. Now I get it. Another victory for code reviews!

Reply:
But wait, I'm using the validity in the `VectorData` created by `Orrify`. Does that not solve the problem?